File: lib/logic/open_skill.dart
import 'dart:math';

/// OpenSkill（Weng-Lin Plackett-Luce）の Dart 実装。
/// openskill.js (MIT) を参考にしています。
class OpenSkillRating {
  final double mu;
  final double sigma;

  const OpenSkillRating({required this.mu, required this.sigma});

  OpenSkillRating copyWith({double? mu, double? sigma}) =>
      OpenSkillRating(mu: mu ?? this.mu, sigma: sigma ?? this.sigma);
}

class OpenSkillConstants {
  static const double defaultMu = 25.0;
  static const double defaultSigma = defaultMu / 3.0;
  static const double beta = defaultMu / 6.0;
  static const double tau = defaultMu / 300.0;
  static const double kappa = 0.0001;
  static const double z = 3.0;
  static const int displayRatingOffset = 1500;

  static OpenSkillRating defaultRating() =>
      const OpenSkillRating(mu: defaultMu, sigma: defaultSigma);
}

class OpenSkill {
  static double ordinal(OpenSkillRating rating) =>
      rating.mu - OpenSkillConstants.z * rating.sigma;

  static int displayRating(OpenSkillRating rating) =>
      (ordinal(rating) + OpenSkillConstants.displayRatingOffset).round();

  /// 旧 ELO のみ保存されているプレイヤーを OpenSkill パラメータへ移行
  static OpenSkillRating fromLegacyRating(int legacyRating) {
    final ordinalValue = legacyRating - OpenSkillConstants.displayRatingOffset;
    return OpenSkillRating(
      mu: ordinalValue + OpenSkillConstants.z * OpenSkillConstants.defaultSigma,
      sigma: OpenSkillConstants.defaultSigma,
    );
  }

  static OpenSkillRating parseStored({
    required dynamic muValue,
    required dynamic sigmaValue,
    required dynamic ratingValue,
  }) {
    if (muValue is num && sigmaValue is num) {
      return OpenSkillRating(mu: muValue.toDouble(), sigma: sigmaValue.toDouble());
    }
    if (ratingValue is num) {
      return fromLegacyRating(ratingValue.round());
    }
    return OpenSkillConstants.defaultRating();
  }

  /// 各チーム 1 プレイヤーの対戦結果から μ・σ を更新する
  static List<List<OpenSkillRating>> rate(
    List<List<OpenSkillRating>> teams,
    List<double> ranks, {
    double tau = OpenSkillConstants.tau,
  }) {
    if (teams.length < 2) return teams;
    if (ranks.length != teams.length) {
      throw ArgumentError('ranks length must match teams length');
    }

    final tauSquared = tau * tau;
    final processed = teams
        .map(
          (team) => team
              .map(
                (p) => p.copyWith(
                  sigma: sqrt(p.sigma * p.sigma + tauSquared),
                ),
              )
              .toList(),
        )
        .toList();

    final indexed = List.generate(
      processed.length,
      (i) => (rank: ranks[i], team: processed[i], index: i),
    )..sort((a, b) => a.rank.compareTo(b.rank));

    final sortedTeams = indexed.map((e) => e.team).toList();
    final sortedRanks = indexed.map((e) => e.rank).toList()..sort();
    final rated = _plackettLuce(sortedTeams, sortedRanks);

    final result = List<List<OpenSkillRating>>.from(processed);
    for (var i = 0; i < indexed.length; i++) {
      result[indexed[i].index] = rated[i];
    }
    return result;
  }

  static List<List<OpenSkillRating>> _plackettLuce(
    List<List<OpenSkillRating>> game,
    List<double> rankInput,
  ) {
    final betaSq = OpenSkillConstants.beta * OpenSkillConstants.beta;
    final teamRatings = _teamRatings(game, rankInput);
    final c = _utilC(teamRatings, betaSq);
    final sumQ = _utilSumQ(teamRatings, c);
    final a = _utilA(teamRatings);

    return List.generate(teamRatings.length, (i) {
      final iTeam = teamRatings[i];
      final iMuOverCe = exp(iTeam.muSum / c);
      var omegaSum = 0.0;
      var deltaSum = 0.0;

      for (var q = 0; q < teamRatings.length; q++) {
        if (teamRatings[q].rank > iTeam.rank) continue;
        final quotient = iMuOverCe / sumQ[q];
        omegaSum += (i == q ? 1 - quotient : -quotient) / a[q];
        deltaSum += (quotient * (1 - quotient)) / a[q];
      }

      final gamma = sqrt(iTeam.sigmaSq) / c;
      final omega = omegaSum * (iTeam.sigmaSq / c);
      final delta = deltaSum * (iTeam.sigmaSq / (c * c)) * gamma;

      return iTeam.team.map((player) {
        final sigmaSq = player.sigma * player.sigma;
        final newMu = player.mu + (sigmaSq / iTeam.sigmaSq) * omega;
        final newSigma = player.sigma *
            sqrt(max(1 - (sigmaSq / iTeam.sigmaSq) * delta, OpenSkillConstants.kappa));
        return OpenSkillRating(mu: newMu, sigma: newSigma);
      }).toList();
    }).toList();
  }

  static List<_TeamRatingData> _teamRatings(
    List<List<OpenSkillRating>> game,
    List<double> rankInput,
  ) {
    final placementRanks = _placementRanks(rankInput);
    return List.generate(game.length, (i) {
      final team = game[i];
      var muSum = 0.0;
      var sigmaSq = 0.0;
      for (final player in team) {
        muSum += player.mu;
        sigmaSq += player.sigma * player.sigma;
      }
      return _TeamRatingData(
        muSum: muSum,
        sigmaSq: sigmaSq,
        team: team,
        rank: placementRanks[i],
      );
    });
  }

  static List<int> _placementRanks(List<double> rankInput) {
    final outRank = List<int>.filled(rankInput.length, 0);
    var s = 0;
    for (var j = 0; j < rankInput.length; j++) {
      if (j > 0 && rankInput[j - 1] < rankInput[j]) {
        s = j;
      }
      outRank[j] = s;
    }
    return outRank;
  }

  static double _utilC(List<_TeamRatingData> teamRatings, double betaSq) {
    var sum = 0.0;
    for (final team in teamRatings) {
      sum += team.sigmaSq + betaSq;
    }
    return sqrt(sum);
  }

  static List<double> _utilSumQ(List<_TeamRatingData> teamRatings, double c) {
    return teamRatings.map((qTeam) {
      var sum = 0.0;
      for (final iTeam in teamRatings) {
        if (iTeam.rank >= qTeam.rank) {
          sum += exp(iTeam.muSum / c);
        }
      }
      return sum;
    }).toList();
  }

  static List<int> _utilA(List<_TeamRatingData> teamRatings) {
    return teamRatings
        .map(
          (iTeam) => teamRatings.where((qTeam) => qTeam.rank == iTeam.rank).length,
        )
        .toList();
  }
}

class _TeamRatingData {
  final double muSum;
  final double sigmaSq;
  final List<OpenSkillRating> team;
  final int rank;

  const _TeamRatingData({
    required this.muSum,
    required this.sigmaSq,
    required this.team,
    required this.rank,
  });
}

File: functions/src/open_skill.ts
export interface OpenSkillRating {
  mu: number;
  sigma: number;
}

export const OpenSkillConstants = {
  defaultMu: 25,
  defaultSigma: 25 / 3,
  beta: 25 / 6,
  tau: 25 / 300,
  kappa: 0.0001,
  z: 3,
  displayRatingOffset: 1500,
};

export function defaultSkillRating(): OpenSkillRating {
  return { mu: OpenSkillConstants.defaultMu, sigma: OpenSkillConstants.defaultSigma };
}

export function ordinal(rating: OpenSkillRating): number {
  return rating.mu - OpenSkillConstants.z * rating.sigma;
}

export function displayRating(rating: OpenSkillRating): number {
  return Math.round(ordinal(rating) + OpenSkillConstants.displayRatingOffset);
}

export function fromLegacyRating(legacyRating: number): OpenSkillRating {
  const ordinalValue = legacyRating - OpenSkillConstants.displayRatingOffset;
  return {
    mu: ordinalValue + OpenSkillConstants.z * OpenSkillConstants.defaultSigma,
    sigma: OpenSkillConstants.defaultSigma,
  };
}

export function parseStoredSkill(data: Record<string, unknown>): OpenSkillRating {
  const mu = data.mu;
  const sigma = data.sigma;
  if (typeof mu === "number" && typeof sigma === "number") {
    return { mu, sigma };
  }
  const rating = data.rating;
  if (typeof rating === "number") {
    return fromLegacyRating(Math.round(rating));
  }
  return defaultSkillRating();
}

interface TeamRatingData {
  muSum: number;
  sigmaSq: number;
  team: OpenSkillRating[];
  rank: number;
}

export function rate(
  teams: OpenSkillRating[][],
  ranks: number[],
  tau = OpenSkillConstants.tau,
): OpenSkillRating[][] {
  if (teams.length < 2) return teams;
  if (ranks.length !== teams.length) {
    throw new Error("ranks length must match teams length");
  }

  const tauSquared = tau * tau;
  const processed = teams.map((team) =>
    team.map((p) => ({
      mu: p.mu,
      sigma: Math.sqrt(p.sigma * p.sigma + tauSquared),
    })),
  );

  const indexed = processed.map((team, index) => ({
    rank: ranks[index],
    team,
    index,
  }));
  indexed.sort((a, b) => a.rank - b.rank);

  const sortedTeams = indexed.map((e) => e.team);
  const sortedRanks = [...indexed.map((e) => e.rank)].sort((a, b) => a - b);
  const rated = plackettLuce(sortedTeams, sortedRanks);

  const result = processed.map((team) => team.map((p) => ({ ...p })));
  for (let i = 0; i < indexed.length; i++) {
    result[indexed[i].index] = rated[i];
  }
  return result;
}

function plackettLuce(game: OpenSkillRating[][], rankInput: number[]): OpenSkillRating[][] {
  const betaSq = OpenSkillConstants.beta * OpenSkillConstants.beta;
  const teamRatings = teamRatingsFromGame(game, rankInput);
  const c = utilC(teamRatings, betaSq);
  const sumQ = utilSumQ(teamRatings, c);
  const a = utilA(teamRatings);

  return teamRatings.map((iTeam, i) => {
    const iMuOverCe = Math.exp(iTeam.muSum / c);
    let omegaSum = 0;
    let deltaSum = 0;

    for (let q = 0; q < teamRatings.length; q++) {
      if (teamRatings[q].rank > iTeam.rank) continue;
      const quotient = iMuOverCe / sumQ[q];
      omegaSum += (i === q ? 1 - quotient : -quotient) / a[q];
      deltaSum += (quotient * (1 - quotient)) / a[q];
    }

    const gamma = Math.sqrt(iTeam.sigmaSq) / c;
    const omega = omegaSum * (iTeam.sigmaSq / c);
    const delta = deltaSum * (iTeam.sigmaSq / (c * c)) * gamma;

    return iTeam.team.map((player) => {
      const sigmaSq = player.sigma * player.sigma;
      const newMu = player.mu + (sigmaSq / iTeam.sigmaSq) * omega;
      const newSigma =
        player.sigma *
        Math.sqrt(Math.max(1 - (sigmaSq / iTeam.sigmaSq) * delta, OpenSkillConstants.kappa));
      return { mu: newMu, sigma: newSigma };
    });
  });
}

function teamRatingsFromGame(
  game: OpenSkillRating[][],
  rankInput: number[],
): TeamRatingData[] {
  const placementRanks = placementRanksFromInput(rankInput);
  return game.map((team, i) => {
    let muSum = 0;
    let sigmaSq = 0;
    for (const player of team) {
      muSum += player.mu;
      sigmaSq += player.sigma * player.sigma;
    }
    return { muSum, sigmaSq, team, rank: placementRanks[i] };
  });
}

function placementRanksFromInput(rankInput: number[]): number[] {
  const outRank = new Array<number>(rankInput.length).fill(0);
  let s = 0;
  for (let j = 0; j < rankInput.length; j++) {
    if (j > 0 && rankInput[j - 1] < rankInput[j]) s = j;
    outRank[j] = s;
  }
  return outRank;
}

function utilC(teamRatings: TeamRatingData[], betaSq: number): number {
  let sum = 0;
  for (const team of teamRatings) sum += team.sigmaSq + betaSq;
  return Math.sqrt(sum);
}

function utilSumQ(teamRatings: TeamRatingData[], c: number): number[] {
  return teamRatings.map((qTeam) => {
    let sum = 0;
    for (const iTeam of teamRatings) {
      if (iTeam.rank >= qTeam.rank) sum += Math.exp(iTeam.muSum / c);
    }
    return sum;
  });
}

function utilA(teamRatings: TeamRatingData[]): number[] {
  return teamRatings.map(
    (iTeam) => teamRatings.filter((qTeam) => qTeam.rank === iTeam.rank).length,
  );
}
